### PR TITLE
Migrate from MdBook to MkDocs

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

We want to have consistency between the Ruff and uv documentation for the upcoming release. We don't love the Ruff docs, but we'd rather have consistency and then work towards improving them both, rather than have two very-different documentation sites that both have weaknesses.

The setup here is simpler than in Ruff as: (1) we don't yet generate any docs from Rust and (2) we don't try to reuse the README in the uv documentation (which adds a lot of complexity in Ruff). So the change here is mostly a 1-to-1 port to MkDocs.

## Test Plan

Screenshot 2024-07-14 at 9 49 15 PM

### CODE REVIEW

File: .github/workflows/publish-docs.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        if: ${{ env.MKDOCS_INSIDERS_SSH_KEY_EXISTS != 'true' }}
+        run: mkdocs build --strict -f mkdocs.public.yml
+      - name: "Deploy to Cloudflare Pages"
+        if: ${{ env.CF_API_TOKEN_EXISTS == 'true' }}
+        uses: cloudflare/wrangler-action@v3.7.0
+        with:
+          apiToken: ${{ secrets.CF_API_TOKEN }}
+          accountId: ${{ secrets.CF_ACCOUNT_ID }}
+          # `github.head_ref` is only set during pull requests and for manual runs or tags we use `main` to deploy to production
+          command: pages deploy site --project-name=uv-docs --branch ${{ github.head_ref || 'main' }} --commit-hash ${GITHUB_SHA}
```
Comment (charliermarsh) [inline .github/workflows/publish-docs.yml:55]: I may create a project here for deployment, or I'll revert back to pages for now.

File: docs/guides/install-python.md [hunk 2]
```diff
@@ -36,7 +36,7 @@ To view available and installed Python versions:
 $ uv python list
 ```
 
-See the [`python list`](../python-versions.html#viewing-available-python-versions) documentation for more details.
+See the [`python list`](../python-versions.md#viewing-available-python-versions) documentation for more details.
```
Comment (charliermarsh) [inline docs/guides/install-python.md:39]: Fixes.

File: docs/first-steps.md [hunk 3]
```diff
@@ -33,7 +33,7 @@ These commands are used to manage Python itself. uv is capable of installing and
 - `uv python list`
 - `uv python find`
 
-See the documentation on [toolchains](./preview/toolchains.md) for more details on getting started.
+See the documentation on [toolchains](./python-versions.md) for more details on getting started.
```
Comment (charliermarsh) [inline docs/first-steps.md:36]: I think?

File: docs/pip/compatibility.md [hunk 4]
```diff
@@ -183,7 +183,7 @@ In other words, uv inverts the default, requiring explicit opt-in to installing
 Python, which can lead to breakages and other complications, and should only be done in limited
 circumstances.
 
-For more, see ["Installing into arbitrary Python environments"](./README.md#installing-into-arbitrary-python-environments).
+For more, see ["Using arbitrary Python environments"](./environments.md#using-arbitrary-python-environments).
```
Comment (charliermarsh) [inline docs/pip/compatibility.md:186]: I think? Will be annoying to keep this in-sync with `PIP_COMPATIBILITY.md`, though maybe the goal is to remove that.

Comment (zanieb) [reply docs/pip/compatibility.md:186]: Yes we'll move `PIP_COMPATIBILITY.md` to here and leave a placeholder with a link,

File: docs/first-steps.md (hunk 3 above)
Comment (zanieb) [reply docs/first-steps.md:36]: Maybe?

```suggestion
See the documentation on [Python version management](./python-versions.md) for more details on getting started.
```

Or maybe the Python guide? I can own fixing this eventually — lots to do in this doc.

File: docs/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+PyYAML==6.0.1
```
Comment (zanieb) [inline docs/requirements.txt:1]: Can we use an `requirements.in` or `pyproject.toml` for these too? Can we dog food here?

**Review by zanieb — APPROVED** [review]

File: docs/requirements.txt (hunk 5 above)
Comment (charliermarsh) [reply docs/requirements.txt:1]: Yeah... I need to figure out what the input requirements were (this is copied from Ruff).

**Pull request merged**